File: apps/web/scripts/import-josaa-cutoffs.mjs
import { createClient } from '@supabase/supabase-js';
import * as fs from 'fs';
import * as path from 'path';
import { parse } from 'csv-parse';
import * as dotenv from 'dotenv';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
dotenv.config({ path: path.resolve(__dirname, '../../../.env.local') });

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY; // Need service role to bypass RLS

if (!supabaseUrl || !supabaseKey) {
  console.error('Missing Supabase credentials');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);

async function run() {
  console.log('Loading CSV...');
  const records = [];
  const csvPath = path.resolve(__dirname, '../../../data/merged_jee_cutoff_2018_2025.csv');
  
  const parser = fs.createReadStream(csvPath).pipe(
    parse({
      columns: true,
      skip_empty_lines: true
    })
  );

  for await (const record of parser) {
    records.push(record);
  }
  
  console.log(`Loaded ${records.length} records from CSV.`);

  // 1. Get or Create a Source
  const { data: sources, error: sourceErr } = await supabase
    .from('sources')
    .select('id')
    .limit(1);
  
  if (sourceErr) throw sourceErr;
  
  let sourceId = sources?.[0]?.id;
  if (!sourceId) {
    const { data: newSource, error: newSourceErr } = await supabase
      .from('sources')
      .insert({
        title: 'JoSAA Cutoffs 2018-2025 Import',
        source_type: 'counselling_authority',
        source_url: 'https://josaa.nic.in',
        academic_year: '2025-2026',
        verification_status: 'published',
        confidence_level: 'A'
      })
      .select()
      .single();
    if (newSourceErr) throw newSourceErr;
    sourceId = newSource.id;
  }
  console.log('Using source_id:', sourceId);

  // 2. Map Colleges
  const { data: dbColleges, error: collErr } = await supabase.from('colleges').select('id, name, slug');
  if (collErr) throw collErr;
  
  const collegeNameMap = new Map();
  for (const coll of dbColleges) {
    collegeNameMap.set(coll.name, coll.id);
  }

  // 3. Delete existing JoSAA cutoffs to avoid duplication
  console.log('Deleting existing JoSAA cutoff records...');
  const { error: delErr } = await supabase
    .from('cutoff_records')
    .delete()
    .eq('counselling_system', 'JoSAA');
  if (delErr) console.log('Delete error (might be fine if none existed):', delErr.message);

  // Process unique colleges and branches
  const uniqueColleges = new Set();
  const uniqueBranches = new Map(); // branchName -> true

  for (const row of records) {
    uniqueColleges.add(row['Institute']);
    const branchKey = `${row['Institute']}___${row['Academic Program Name']}`;
    uniqueBranches.set(branchKey, {
      collegeName: row['Institute'],
      branchName: row['Academic Program Name']
    });
  }

  console.log(`Found ${uniqueColleges.size} unique colleges and ${uniqueBranches.size} unique branches in CSV.`);

  // Insert missing colleges
  for (const institute of uniqueColleges) {
    if (!collegeNameMap.has(institute)) {
      console.log(`Inserting missing college: ${institute}`);
      const slug = institute.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/(^-|-$)/g, '');
      const { data: newColl, error: insCollErr } = await supabase
        .from('colleges')
        .insert({
          name: institute,
          slug: slug,
          ownership: 'GOVERNMENT',
          city: 'Unknown',
          state: 'Unknown',
          is_published: true
        })
        .select()
        .single();
      if (insCollErr) {
        console.error('Error inserting college', institute, insCollErr.message);
      } else {
        collegeNameMap.set(institute, newColl.id);
      }
    }
  }

  // Fetch all branches
  const { data: dbBranches, error: branchErr } = await supabase.from('college_branches').select('id, college_id, name');
  if (branchErr) throw branchErr;

  const branchMap = new Map(); // key: "collegeId___branchName" -> branchId
  for (const b of dbBranches) {
    branchMap.set(`${b.college_id}___${b.name}`, b.id);
  }

  // Insert missing branches
  for (const [key, branchInfo] of uniqueBranches.entries()) {
    const collegeId = collegeNameMap.get(branchInfo.collegeName);
    const mapKey = `${collegeId}___${branchInfo.branchName}`;
    if (!branchMap.has(mapKey)) {
      // Create branch
      const { data: newBranch, error: newBranchErr } = await supabase
        .from('college_branches')
        .insert({
          college_id: collegeId,
          name: branchInfo.branchName,
          degree: 'B.Tech',
          duration_years: 4,
          source_id: sourceId,
          verification_status: 'published',
          confidence_level: 'A'
        })
        .select()
        .single();
      
      if (newBranchErr) {
        console.error('Error inserting branch', branchInfo.branchName, newBranchErr.message);
      } else {
        branchMap.set(mapKey, newBranch.id);
      }
    }
  }

  // 4. Batch insert cutoffs
  console.log('Inserting cutoff records...');
  
  const cutoffsToInsert = [];
  for (const row of records) {
    const collegeId = collegeNameMap.get(row['Institute']);
    const mapKey = `${collegeId}___${row['Academic Program Name']}`;
    const branchId = branchMap.get(mapKey);

    if (!branchId) {
      console.warn('Missing branch ID for', row['Academic Program Name']);
      continue;
    }

    const isIIT = row['Institute'].includes('Indian Institute of Technology') && !row['Institute'].includes('Information Technology');
    const examName = isIIT ? 'JEE Advanced' : 'JEE Main';

    const opRank = parseFloat(row['Opening Rank']);
    const clRank = parseFloat(row['Closing Rank']);

    cutoffsToInsert.push({
      college_id: collegeId,
      branch_id: branchId,
      exam: examName,
      year: parseInt(row['Year']),
      admission_year: parseInt(row['Year']),
      counselling_system: 'JoSAA',
      round: row['Round'],
      category: row['Seat Type'],
      quota: row['Quota'],
      gender_pool: row['Gender'],
      opening_rank: isNaN(opRank) ? null : opRank,
      closing_rank: isNaN(clRank) ? 0 : clRank,
      source_id: sourceId,
      verification_status: 'published',
      publication_status: 'published',
      confidence_level: 'A'
    });
  }

  const batchSize = 1000;
  for (let i = 0; i < cutoffsToInsert.length; i += batchSize) {
    const batch = cutoffsToInsert.slice(i, i + batchSize);
    const { error: insertErr } = await supabase
      .from('cutoff_records')
      .insert(batch);
    if (insertErr) {
      console.error('Error inserting batch at index', i, insertErr.message);
    } else {
      console.log(`Inserted batch ${i} to ${i + batchSize}`);
    }
  }

  console.log('Done!');
}

run().catch(console.error);

File: apps/data-pipeline/scripts/import_josaa_cutoffs.py
import argparse
import csv
import logging
import os
import time
from pathlib import Path

import httpx

logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")

def load_env(env_path: Path):
    if not env_path.exists():
        return
    with open(env_path, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith("#"):
                key, val = line.split("=", 1)
                if key not in os.environ:
                    os.environ[key] = val.strip("'\"")

class SupabaseClient:
    def __init__(self, url: str, key: str):
        self.url = f"{url}/rest/v1"
        self.headers = {
            "apikey": key,
            "Authorization": f"Bearer {key}",
            "Content-Type": "application/json",
            "Prefer": "return=representation"
        }
        self.client = httpx.Client(base_url=self.url, headers=self.headers, timeout=60.0)

    def get_colleges(self) -> list:
        r = self.client.get("/colleges?select=id,name")
        r.raise_for_status()
        return r.json()
        
    def get_college_branches(self) -> list:
        r = self.client.get("/college_branches?select=id,college_id,name")
        r.raise_for_status()
        return r.json()
        
    def create_college(self, name: str) -> str:
        slug = name.lower().replace(" ", "-").replace(",", "").replace("(", "").replace(")", "")
        r = self.client.post("/colleges", json=[{
            "name": name,
            "slug": slug,
            "is_published": True,
            "verification_status": "published"
        }], headers={"Prefer": "return=representation"})
        r.raise_for_status()
        return r.json()[0]["id"]
        
    def create_branch(self, college_id: str, name: str) -> str:
        r = self.client.post("/college_branches", json=[{
            "college_id": college_id,
            "name": name,
            "verification_status": "published",
            "confidence_level": "A"
        }], headers={"Prefer": "return=representation"})
        r.raise_for_status()
        return r.json()[0]["id"]
        
    def delete_existing_cutoffs(self):
        # We need to delete in batches or use a fast delete if we're replacing
        # For safety and speed, we will just delete where counselling_system = JoSAA
        r = self.client.delete("/cutoff_records?counselling_system=eq.JoSAA")
        r.raise_for_status()
        logging.info("Deleted existing JoSAA cutoffs.")

    def bulk_insert_cutoffs(self, payload: list[dict]):
        if not payload:
            return
        r = self.client.post("/cutoff_records", json=payload, headers={"Prefer": "return=minimal"})
        if r.status_code not in (200, 201, 204):
            logging.error(f"Failed to insert batch: {r.text}")
            r.raise_for_status()

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--csv", required=True, help="Path to CSV file")
    args = parser.parse_args()

    # Calculate absolute path for .env.local
    current_dir = Path(__file__).parent.resolve()
    env_path = current_dir.parent.parent / "web" / ".env.local"
    
    load_env(env_path)

    url = os.environ.get("NEXT_PUBLIC_SUPABASE_URL")
    key = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")

    if not url or not key:
        logging.error(f"Missing NEXT_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY (looked in {env_path})")
        return

    client = SupabaseClient(url, key)
    
    logging.info("Fetching existing colleges and branches...")
    existing_colleges = client.get_colleges()
    college_name_to_id = {c["name"]: c["id"] for c in existing_colleges}
    
    existing_branches = client.get_college_branches()
    # map (college_id, branch_name) to branch_id
    branch_map = {(b["college_id"], b["name"]): b["id"] for b in existing_branches}
    
    logging.info("Deleting old JoSAA cutoffs...")
    client.delete_existing_cutoffs()
    
    logging.info("Reading CSV...")
    
    batch = []
    batch_size = 3000
    total_inserted = 0
    
    with open(args.csv, "r", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            college_name = row["Institute"].strip()
            branch_name = row["Academic Program Name"].strip()
            quota = row["Quota"].strip()
            category = row["Seat Type"].strip()
            gender_pool = row["Gender"].strip()
            
            try:
                opening_rank = float(row["Opening Rank"]) if row["Opening Rank"] else None
                closing_rank = float(row["Closing Rank"]) if row["Closing Rank"] else None
            except ValueError:
                continue
                
            round_val = row["Round"].strip()
            
            try:
                year = int(row["Year"])
            except ValueError:
                continue
            
            if closing_rank is None:
                continue
                
            # Get or create college
            if college_name not in college_name_to_id:
                logging.info(f"Creating missing college: {college_name}")
                cid = client.create_college(college_name)
                college_name_to_id[college_name] = cid
            
            college_id = college_name_to_id[college_name]
            
            # Get or create branch
            branch_key = (college_id, branch_name)
            if branch_key not in branch_map:
                logging.info(f"Creating missing branch: {branch_name} in {college_name}")
                bid = client.create_branch(college_id, branch_name)
                branch_map[branch_key] = bid
                
            branch_id = branch_map[branch_key]
            
            exam = "JEE Advanced" if "Indian Institute of Technology" in college_name else "JEE Main"
            
            batch.append({
                "college_id": college_id,
                "college_branch_id": branch_id,
                "exam": exam,
                "counselling_system": "JoSAA",
                "admission_year": year,
                "round": round_val,
                "category": category,
                "quota": quota,
                "gender_pool": gender_pool,
                "opening_rank": opening_rank,
                "closing_rank": closing_rank,
                "verification_status": "published",
                "publication_status": "published",
                "source_id": f"josaa_bulk_{year}"
            })
            
            if len(batch) >= batch_size:
                client.bulk_insert_cutoffs(batch)
                total_inserted += len(batch)
                logging.info(f"Inserted {total_inserted} records...")
                batch = []
                time.sleep(0.2)
                
        if batch:
            client.bulk_insert_cutoffs(batch)
            total_inserted += len(batch)
            logging.info(f"Inserted {total_inserted} records...")
            
    logging.info("Import complete.")

if __name__ == "__main__":
    main()
